### ISSUE TITLE
Feature Request: Matrix measurements

### ISSUE THREAD

**Issue by heca-project** [sharkdp/hyperfine#253]
If I want to benchmark my program with one variable command argument, I can use `--parameter-list`. On the other hand, if I want to benchmark four, five, six variables, it can get tedious.

It would be nice if I'd be able to run:

```
hyperfine 'arg {a1} {a2} {a3}' --parameter-list a1 1,2,3 --parameter-list a2 a,b,c --parameter-list a3 Q,W,E
```

**Comment by sharkdp** [comment]
Thank you very much for your feedback.

I agree that this would be a useful feature to have.

Implementation-wise, we could probably take a first step by converting the "matrix" to a flat list. This would mean that we would *not* add special support in the terminal output, the export formats, etc. It would simply be a list of benchmarks. 27 in your example:
```
1 a Q
1 a W
1 a E
1 b Q
...
3 c E
```

**Comment by albins** [comment]
Expanding a bit on this, I think there would be at least two modes of lists, zip (i.e. pair up the two variable parameters) and cross product (all combinations of values). It's not a given that you would want to support both I guess?

**Comment by sharkdp** [comment]
Thank you for the feedback. I would go with the cross-product version only, I think.

I think the zipped-version is much less common and (while not pretty), it probably can be simulated in most cases:
```bash
hyperfine --parameter-list args "--foo 1 --bar a","--foo 2 --bar b" "my-command {args}"
```

**Comment by Sarcasm** [comment]
Also interested by this feature, in my case, I was expecting a cross product:

    hyperfine -r 2 --show-output -L mode a,b -P x 0 2 -P y 0 2 'echo mode={mode} x={x} y={y}'

Zipping would also be nice, but could be added separately with a new syntax, e.g.: `--zip-parameter x,y`?

**Comment by albins** [comment]
I have been thinking about this. My own use case is to race N implementations of SMT solvers against each other for sets of benchmarks (input files). If N=2 and I am A/B testing (which would be my typical use case during development), this would in principle be equivalent to a shell for loop executing a regular "compare two" input to Hyperfine with the benchmark files coming from the loop.

However, the situation becomes a bit more interesting if I were to compare three or more implementations; then I would be interested in a comparison between the best and the second-best.

The main design challenge I see to getting something useful out the door is to let the user somehow control how the results are generated. For example, if I have parameters (algoA,algoB), (fileA,fileB), (optionA, optionB) I would like to predictably have an output like:

```
fileA algoA optionA
fileA algoA optionB
fileA algoB optionA
...
fileB algoB optionB
```

What are your thoughts, @sharkdp?

**Comment by wchargin** [comment]
I’ve sent a pull request for this over at #318, which implements a cross
product. Feel free to try it out and comment on the PR if you have any
feedback. Thanks!

**Comment by sharkdp** [comment]
This is now supported thanks to #318 by @wchargin!

**Comment by sharkdp** [comment]
Released in [hyperfine v1.11.0](https://github.com/sharkdp/hyperfine/releases/tag/v1.11.0).
